File: HasingAndArrays/ProductArrayExceptSelf/238-product-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        nLen = len(nums)
        output = [1] * nLen

        # prefix calculations
        for i in range(1, nLen):
            output[i] = output[i - 1] * nums[i - 1]

        # suffix caculations with helper variable that
        # accumulates the products up to the current index (exclusive)
        cumul = 1
        for i in range(nLen - 2, -1, -1):
            cumul *= nums[i + 1]
            output[i] *= cumul

        return output

File: HasingAndArrays/ProductArrayExceptSelf/238-product-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int totalProduct = 1;
        int zeroIdx = -1;

        for (int i = 0; i < nums.length; i++) {
            if (nums[i] == 0 && zeroIdx != -1) {
                zeroIdx = i;
                continue;
            } else if (nums[i] == 0) {
                return new int[nums.length];
            }

            totalProduct *= nums[i];
        }

        if (zeroIdx != -1) {
            int[] result = new int[nums.length];
            result[zeroIdx] = totalProduct;
            return result;
        }

        for (int i = 0; i < nums.length; i++) {
            nums[i] = totalProduct / nums[i];
        }

        return nums;
    }
}  
